strawberry/field.py

```diff
@@ -23,129 +23,289 @@
+    is_basic_field: bool = False
+
+    def __post_init__(self):
+        # any errors due to invalid implementation
+        # should be reported here,
+        # because any evolve / dataclasses.replace will call __post_init__.
+        if isinstance(self.base_resolver, StrawberryResolver):
+            for argument in self.base_resolver.arguments:
+                if isinstance(argument.type_annotation.annotation, str):
+                    continue
+                elif isinstance(argument.type, StrawberryUnion):
+                    raise InvalidFieldArgument(
+                        self.base_resolver.name,
+                        argument.python_name,
+                        "Union",
+                    )
 
-        self.permission_classes: List[Type[BasePermission]] = list(permission_classes)
-        self.directives = directives
+                elif getattr(argument.type, "_type_definition", False):
+                    if argument.type._type_definition.is_interface:  # type: ignore
+                        raise InvalidFieldArgument(
+                            self.base_resolver.name,
+                            argument.python_name,
+                            "Interface",
+                        )
+        # checks that only applied after origin has been determined.
+        # this would normally be checked after StrawberryField.evolve has been called.
+        if self.origin:
+            # Check that the field type is not Private
+            if is_private(self.type_annotation.resolve()):
+                raise PrivateStrawberryFieldError(
+                    self.python_name, self.origin.__name__
+                )
+            # Check that default is not set if a resolver is defined
+            if (
+                self.dataclass_options.default is not UNSET
+                and self.base_resolver is not None
+            ):
+                raise FieldWithResolverAndDefaultValueError(
+                    self.python_name, self.origin.__name__
+                )
+            # Check that default_factory is not set if a resolver is defined
+            # Note: using getattr because of this issue:
+            # https://github.com/python/mypy/issues/6910
+            if (
+                self.dataclass_options.default_factory is not UNSET  # noqa
+                and self.base_resolver is not None
+            ):
+                raise FieldWithResolverAndDefaultFactoryError(
+                    self.python_name, self.origin.__name__
+                )
 
-        self.deprecation_reason = deprecation_reason
+            assert_message = (
+                "Field must have a name by the time the schema is generated"
+            )
+            assert self.python_name is not None, assert_message
 
-    def __call__(self, resolver: _RESOLVER_TYPE) -> "StrawberryField":
-        """Add a resolver to the field"""
+    @classmethod
+    def create(
```

Comment:
why a class method and not just a constructor? :)

Reply:
lol I though this would allows to modify the `dataclass __init__` along the way didn't though to call `super().__init__(...)`

Reply:
actually I was right, you cannot assign to a frozen dataclass in __init__ see [this SO](https://stackoverflow.com/questions/57893902/how-can-i-set-an-attribute-in-a-frozen-dataclass-custom-init-method)